backend/provisioner/dev/dev_provisioner.go

```diff
@@ -0,0 +1,187 @@
+package dev
+
+import (
+	"context"
+	"math/rand/v2"
+	"strconv"
+	"sync/atomic"
+
+	"connectrpc.com/connect"
+	"github.com/XSAM/otelsql"
+
+	ftlv1 "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1beta1/provisioner"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1beta1/provisioner/provisionerconnect"
+	"github.com/TBD54566975/ftl/internal/dev"
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+type task struct {
+	steps []*step
+}
+
+func (t *task) Done() bool {
+	for _, step := range t.steps {
+		if !step.done.Load() {
+			return false
+		}
+	}
+	return true
+}
+
+type step struct {
+	resource *provisioner.Resource
+	err      error
+	done     atomic.Bool
+}
+
+// Provisioner is a provisioner for running FTL locally
+type Provisioner struct {
+	running     map[string]*task
+	postgresDSN string
+
+	postgresImage string
+	postgresPort  int
+}
+
+func NewProvisioner(postgresImage string, postgresPort int) *Provisioner {
+	return &Provisioner{
+		postgresImage: postgresImage,
+		postgresPort:  postgresPort,
+	}
+}
+
+var _ provisionerconnect.ProvisionerPluginServiceClient = (*Provisioner)(nil)
+
+func (d *Provisioner) Ping(context.Context, *connect.Request[ftlv1.PingRequest]) (*connect.Response[ftlv1.PingResponse], error) {
+	return &connect.Response[ftlv1.PingResponse]{}, nil
+}
+
+func (d *Provisioner) Plan(context.Context, *connect.Request[provisioner.PlanRequest]) (*connect.Response[provisioner.PlanResponse], error) {
+	panic("unimplemented")
+}
+
+func (d *Provisioner) Provision(ctx context.Context, req *connect.Request[provisioner.ProvisionRequest]) (*connect.Response[provisioner.ProvisionResponse], error) {
+	if d.running == nil {
+		d.running = map[string]*task{}
+	}
+
+	previous := map[string]*provisioner.Resource{}
+	for _, r := range req.Msg.ExistingResources {
+		previous[r.ResourceId] = r
+	}
+
+	task := &task{}
+	for _, r := range req.Msg.DesiredResources {
+		if _, ok := previous[r.Resource.ResourceId]; !ok {
+			switch tr := r.Resource.Resource.(type) {
+			case *provisioner.Resource_Postgres:
+				step := &step{resource: r.Resource}
+				task.steps = append(task.steps, step)
+
+				d.provisionPostgres(ctx, tr, req.Msg.Module, r.Resource.ResourceId, step)
+			default:
```

Comment:
Return an error here? The gRPC status code INVALID_ARGUMENT might be appropriate.